llarp/messages/relay_status.cpp

```diff
@@ -0,0 +1,295 @@
+    {
+      return false;
+    }
+
+    QueueSendMessage(router, nextHop, message);
+    return true; // can't guarantee delivery here, as far as we know it's fine
+  }
+
+  bool
+  LR_StatusMessage::AddFrame(const SharedSecret& pathKey, uint64_t status)
+  {
+    frames[7] = frames[6];
+    frames[6] = frames[5];
+    frames[5] = frames[4];
+    frames[4] = frames[3];
+    frames[3] = frames[2];
+    frames[2] = frames[1];
+    frames[1] = frames[0];
+
+    auto& frame = frames[0];
+
+    frame.Randomize();
+
+    LR_StatusRecord record;
+
+    record.status    = status;
+    record.version     = LLARP_PROTO_VERSION;
+
+    llarp_buffer_t buf(frame.data(), frame.size());
+    buf.cur = buf.base + EncryptedFrameOverheadSize;
+    // encode record
+    if(!record.BEncode(&buf))
+    {
+      // failed to encode?
+      LogError(Name(), " Failed to generate Status Record");
+      DumpBuffer(buf);
+      return false;
+    }
+    // use ephemeral keypair for frame
+    if(!frame.DoEncrypt(pathKey, true))
+    {
+      LogError(Name(), " Failed to encrypt LRSR");
+      DumpBuffer(buf);
+      return false;
+    }
+
+    return true;
+  }
+
+  void
+  LR_StatusMessage::QueueSendMessage(AbstractRouter *router, const RouterID nextHop,
+                   std::shared_ptr< LR_StatusMessage > msg)
+  {
+    auto func = std::bind(&LR_StatusMessage::SendMessage, router, nextHop, msg); 
+    router->pathContext().logic()->queue_func(func);
+  }
+
+  void
+  LR_StatusMessage::SendMessage(AbstractRouter *router, const RouterID nextHop,
+                                std::shared_ptr< LR_StatusMessage > msg)
+  {
+    llarp::LogDebug("Attempting to send LR_Status message to (", nextHop, ")");
+    if(router->HasSessionTo(nextHop))
+    {
+      if (router->SendToOrQueue(nextHop, msg.get()))
+      {
+        return;
+      }
+      llarp::LogError("Sending LR_Status message, SendToOrQueue to ", nextHop, " failed");
+      return;
+    }
+    llarp::LogError("Sending LR_Status message, but no connection to previous hop (", nextHop, ")");
+    return;
+  }
+
+  bool
+  LR_StatusRecord::BEncode(llarp_buffer_t* buf) const
+  {
+    if(!bencode_start_dict(buf))
+      return false;
+
+    if(!BEncodeWriteDictInt("s", status, buf))
+      return false;
+    if(!bencode_write_version_entry(buf))
+      return false;
+
+    return bencode_end(buf);
+  }
+
+  bool
+  LR_StatusRecord::OnKey(llarp_buffer_t* buffer, llarp_buffer_t* key)
+  {
+    if(!key)
+      return true;
+
+    bool read = false;
+
+    if(!BEncodeMaybeReadDictInt("s", status, read, *key, buffer))
+      return false;
+    if(!BEncodeMaybeReadVersion("v", version, LLARP_PROTO_VERSION, read, *key,
+                                buffer))
+      return false;
+
+    return read;
+  }
+
+  bool
+  LR_StatusRecord::BDecode(llarp_buffer_t* buf)
+  {
+    using namespace std::placeholders;
+    return bencode_read_dict(std::bind(&LR_StatusRecord::OnKey, this, _1, _2),
```

Comment:
```c++
return bencode_read_dict(util::memFn(&LR_StatusRecord::OnKey, this), buf);
```

Reply:
will do; I don't think that bit was merged yet when I was working on this and I forgot it was added/changed.

Reply:
addressed